applications/SeismicApplication/custom_elements/fiber_beam_column_element_3D2N.cpp

```diff
@@ -0,0 +1,556 @@
+
+    reference_coordinates[0] = GetGeometry()[0].X0();
+    reference_coordinates[1] = GetGeometry()[0].Y0();
+    reference_coordinates[2] = GetGeometry()[0].Z0();
+    reference_coordinates[3] = GetGeometry()[1].X0();
+    reference_coordinates[4] = GetGeometry()[1].Y0();
+    reference_coordinates[5] = GetGeometry()[1].Z0();
+
+    for (unsigned int i = 0; i < msDimension; ++i) {
+        direction_vector_x[i] =
+            (reference_coordinates[i + msDimension] - reference_coordinates[i]);
+    }
+
+    array_1d<double, msDimension> global_z = ZeroVector(msDimension);
+    global_z[2] = 1.0;
+    array_1d<double, msDimension> v2 = ZeroVector(msDimension);
+    array_1d<double, msDimension> v3 = ZeroVector(msDimension);
+
+    double vector_norm;
+    vector_norm = MathUtils<double>::Norm(direction_vector_x);
+    if (vector_norm > numerical_limit) {
+        direction_vector_x /= vector_norm;
+    }
+
+    if (std::abs(direction_vector_x[2] - 1.00) < numerical_limit) {
+            v2[1] = 1.0;
+            v3[0] = -1.0;
+    } else if (std::abs(direction_vector_x[2] + 1.00) < numerical_limit) {
+        v2[1] = 1.0;
+        v3[0] = 1.0;
+    } else {
+        MathUtils<double>::UnitCrossProduct(v2, global_z, direction_vector_x);
+        MathUtils<double>::UnitCrossProduct(v3, direction_vector_x, v2);
+    }
+
+    for (unsigned int i = 0; i < msDimension; ++i) {
+        local_cs(i, 0) = direction_vector_x[i];
+        local_cs(i, 1) = v2[i];
+        local_cs(i, 2) = v3[i];
+    }
+
+    return local_cs;
+    KRATOS_CATCH("")
+}
+
+double FiberBeamColumnElement3D2N::CalculateReferenceLength() const
+{
+    KRATOS_TRY
+
+    const array_1d<double, 3> delta_pos =
+        GetGeometry()[1].GetInitialPosition().Coordinates() -
+        GetGeometry()[0].GetInitialPosition().Coordinates();
+
+    return std::sqrt((delta_pos[0] * delta_pos[0]) +
+                     (delta_pos[1] * delta_pos[1]));
+
+    KRATOS_CATCH("")
+}
+
+int FiberBeamColumnElement3D2N::Check(const ProcessInfo& rCurrentProcessInfo)
+{
+    KRATOS_TRY
+
+    KRATOS_ERROR_IF(this->Id() < 1) <<"FiberBeamColumnElement3D2N found with Id 0 or negative" << std::endl;
+
+    KRATOS_ERROR_IF(this->GetGeometry().Area() <= 0) << "On FiberBeamColumnElement3D2N -> "
+        << this->Id() <<  "; Area cannot be less than or equal to 0" << std::endl;
+
+    return 0;
+
+    KRATOS_CATCH("");
+}
+
+///@}
+///@name Access
+///@{
+
+
+///@}
+///@name Inquiry
+///@{
+
+
+///@}
+///@name Input and output
+///@{
+
+std::string FiberBeamColumnElement3D2N::Info() const {
+    std::stringstream buffer;
+    buffer << "FiberBeamColumnElement3D2N #" << Id();
+    return buffer.str();
+}
+
+void FiberBeamColumnElement3D2N::PrintInfo(std::ostream& rOStream) const {
+    rOStream << "FiberBeamColumnElement3D2N #" << Id();
+}
+
+void FiberBeamColumnElement3D2N::PrintData(std::ostream& rOStream) const {
+    pGetGeometry()->PrintData(rOStream);
+}
+
```

Comment:
What the main difference with the beam element already existing?

Reply:
Differences are:
* Geometry is a `line_gauss_lobatto_3d_2`, similar but another integration rule
* The integration points are sections, which are discretized into fibers
* After every non-linear iteration, a loop of element equilibrium is performed
* The stiffness matrix is the inverse of the integral of section flexibility matrices

Reply:
Ok, can we look for an approach in order to reduce code duplication?

Reply:
Technically, some methods are exactly the same, because the global dofs are the same. The local however are different. It is only one axial force and two biaxial moments in this element. So in principle maybe they can share a base class?

Reply:
Indeed

Reply:
I suggest to wait with this for two reasons:
- Many methods in the structural elements are duplicated, I was talking to @KlausBSautter that we should do a proper refactor soon
- @KlausBSautter has implemented a new beam element, which still needs to be merged, it will replace completely the current one